### PR TITLE
feat: mcopy

### PR DESCRIPTION
**Pull request by obatirou**
## Pull Request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build-related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

Resolves: #834
## What is the new behavior?

Implement `mcopy`

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

- - -
This change is [Reviewable](https://reviewable.io/reviews/kkrt-labs/kakarot-ssj/866)

### CODE REVIEW

File: crates/evm/src/instructions/memory_operations.cairo [hunk 1]
```diff
@@ -248,13 +249,35 @@ impl MemoryOperation of MemoryOperationTrait {
...
+        let dest_offset = self.stack.pop_usize()?;
+
+        if size == 0 {
+            return Result::Ok(());
+        }
+
+        let memory_expansion = gas::memory_expansion(
+            self.memory.size(), max(dest_offset, source_offset) + size
+        );
+        self.charge_gas(gas::VERYLOW + memory_expansion.expansion_cost)?;
```
Comment (enitrat) [inline crates/evm/src/instructions/memory_operations.cairo:268]: gas should be charged even if the size is 0. 
generally we follow the

// 1. STACK OPERATIONS
// 2. GAS OPERATIONS
// 3. OPCODE OPERATIONS

File: crates/evm/src/instructions/memory_operations.cairo [hunk 2]
```diff
@@ -915,4 +938,45 @@ mod tests {
...
+        assert(vm.stack.is_empty(), 'stack should be empty');
+        assert(vm.memory.size() == dest_offset + size, 'memory size error');
+        i = 0;
+        for element in values
+            .span() {
+                let stored_word = vm.memory.load(dest_offset + 0x20 * i);
+                assert(stored_word == (*element).into(), 'mcopy failed');
+                i += 1;
+            };
+    }
```
Comment (enitrat) [inline crates/evm/src/instructions/memory_operations.cairo:981]: given that the gas charged is constant, we can trivially verify it here

**Review by enitrat — COMMENTED** [review]
great !

File: crates/evm/src/instructions/memory_operations.cairo (hunk 2 above)
Comment (obatirou) [reply crates/evm/src/instructions/memory_operations.cairo:988]: done with [0f169fe](https://github.com/kkrt-labs/kakarot-ssj/pull/866/commits/0f169fe40b99c93c237f4f8abf807f0c180adb81)

**Pull request merged**